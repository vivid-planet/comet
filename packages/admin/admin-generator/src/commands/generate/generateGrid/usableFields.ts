/* eslint-disable @typescript-eslint/no-explicit-any */

type Prev = [never, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9];

type GqlLeaves<T, FollowArrays extends boolean = false, Depth extends number = 5> = [Depth] extends [never]
    ? never
    : T extends any
      ? T extends Array<infer ArrayType>
          ? FollowArrays extends true
              ? GqlLeaves<ArrayType, FollowArrays, Prev[Depth]>
              : never
          : "__typename" extends keyof T
            ? {
                  [K in keyof T as K extends "__typename" ? never : K]-?: GqlLeaves<T[K], FollowArrays, Prev[Depth]>;
              }
            : never
      : never;
type IfExplicitAny<T, Y, N> = T extends never ? Y : N;

// UsableFields steps into objects containing __typename but does not include those objects themselves
type FieldNames<T> = {
    [K in keyof T]: `${Exclude<K, symbol>}${FieldNames<T[K]> extends never ? "" : `.${FieldNames<T[K]>}`}`;
}[keyof T];
export type UsableFields<T, FollowArrays extends boolean = false> = IfExplicitAny<T, any, FieldNames<GqlLeaves<T, FollowArrays>>>;

// UsableFormFields steps into objects containing __typename and does include those objects themselves
type FormFieldNames<T> = {
    [K in keyof T]: `${Exclude<K, symbol>}` | (FieldNames<T[K]> extends never ? never : `${Exclude<K, symbol>}.${FieldNames<T[K]>}`);
}[keyof T];
export type UsableFormFields<T> = IfExplicitAny<T, any, FormFieldNames<GqlLeaves<T, false>>>;
