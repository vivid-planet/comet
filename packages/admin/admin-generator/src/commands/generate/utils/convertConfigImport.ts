import { type Imports } from "./generateImportsCode";

// generated code is one level below config, this function converts relative imports to go one level up
export function convertConfigImport(imprt: { name: string; import: string }): Imports[0] {
    let importPath = imprt.import as string;
    if (importPath.startsWith("../")) {
        importPath = `../${importPath}`;
    } else if (importPath.startsWith("./")) {
        importPath = `.${importPath}`;
    }
    return { name: imprt.name, importPath };
}
