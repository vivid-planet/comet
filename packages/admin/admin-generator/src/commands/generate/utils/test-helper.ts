import { Project, type SourceFile } from "ts-morph";

export function parseSource(source: string): SourceFile {
    const project = new Project({
        tsConfigFilePath: "./tsconfig.json",
        skipAddingFilesFromTsConfig: true,
    });
    return project.createSourceFile("test.ts", source);
}
