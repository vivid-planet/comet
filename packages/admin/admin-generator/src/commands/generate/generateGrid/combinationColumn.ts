import { type GridColDef } from "@comet/admin";
import { type ComponentProps } from "react";
import { type FormattedNumber } from "react-intl";

import { type BaseColumnConfig } from "../generate-command";
import { getFormattedMessageNode } from "../utils/intl";

type AbstractField<FieldName extends string> = {
    field: FieldName;
    emptyValue?: string;
};

type TextField<FieldName extends string> = AbstractField<FieldName> & {
    type: "text";
};

type StaticText = {
    type: "static";
    text: string;
};

type FormattedNumberPropsForNumberField = Omit<ComponentProps<typeof FormattedNumber>, "value" | "children">;

type NumberField<FieldName extends string> = AbstractField<FieldName> &
    FormattedNumberPropsForNumberField & {
        type: "number";
        decimals?: number;
    };

type StaticSelectField<FieldName extends string> = AbstractField<FieldName> & {
    type: "staticSelect";
    values: Array<
        | string
        | {
              value: string | number | boolean;
              label: string;
          }
    >;
};

type FormattedMessage<FieldName extends string> = {
    type: "formattedMessage";
    message: string;
    valueFields: Record<string, Field<FieldName>>;
};

type Field<FieldName extends string> =
    | StaticText
    | FieldName
    | TextField<FieldName>
    | NumberField<FieldName>
    | StaticSelectField<FieldName>
    | FormattedMessage<FieldName>;

export type GridCombinationColumnConfig<FieldName extends string> = {
    type: "combination";
    name: string;
    primaryText?: Field<FieldName>;
    secondaryText?: Field<FieldName>;
} & BaseColumnConfig &
    Pick<GridColDef, "sortBy">;

type CellContent = {
    textContent: string;
    variableDefinitions?: string[];
};

const getTextForCellContent = (textConfig: Field<string>, messageIdPrefix: string): CellContent => {
    if (typeof textConfig !== "string" && textConfig.type === "static") {
        return {
            textContent: getFormattedMessageNode(messageIdPrefix, textConfig.text),
        };
    }

    if (typeof textConfig !== "string" && textConfig.type === "formattedMessage") {
        const variableDefinitions: string[] = [];

        const values = Object.entries(textConfig.valueFields)
            .map(([key, value]) => {
                const { textContent, variableDefinitions: cellVariableDefinitions } = getTextForCellContent(value, `${messageIdPrefix}.${key}`);

                if (cellVariableDefinitions?.length) {
                    variableDefinitions.push(...cellVariableDefinitions);
                }

                return `${key}: ${textContent}`;
            })
            .join(", ");

        return {
            textContent: getFormattedMessageNode(messageIdPrefix, textConfig.message, `{${values}}`),
            variableDefinitions,
        };
    }

    const emptyText =
        typeof textConfig !== "string" && "emptyValue" in textConfig
            ? getFormattedMessageNode(`${messageIdPrefix}.empty`, textConfig.emptyValue)
            : "'-'";

    const fieldName = typeof textConfig === "string" ? textConfig : textConfig.field;
    const rowValue = `row.${fieldName.replace(/\./g, "?.")}`;
    const stringTextContent = `${rowValue} ?? ${emptyText}`;

    if (typeof textConfig === "string") {
        return {
            textContent: stringTextContent,
        };
    }

    if (textConfig.type === "number") {
        const { type, decimals: decimalsConfigValue, field, emptyValue, ...configForFormattedNumberProps } = textConfig;

        const hasCurrency = Boolean(textConfig.currency);
        const hasUnit = Boolean(textConfig.unit);

        const defaultDecimalsProp = hasCurrency && decimalsConfigValue === undefined ? 2 : decimalsConfigValue;
        let defaultStyleProp: string | undefined = undefined;

        if (hasCurrency) {
            defaultStyleProp = '"currency"';
        } else if (hasUnit) {
            defaultStyleProp = '"unit"';
        }

        const formattedNumberProps: Record<string, unknown> = {
            value: `{${rowValue}}`,
            minimumFractionDigits: typeof defaultDecimalsProp !== "undefined" ? `{${defaultDecimalsProp}}` : undefined,
            maximumFractionDigits: typeof defaultDecimalsProp !== "undefined" ? `{${defaultDecimalsProp}}` : undefined,
            style: typeof defaultStyleProp !== "undefined" ? defaultStyleProp : undefined,
        };

        Object.entries(configForFormattedNumberProps).forEach(([key, value]) => {
            if (typeof value === "string") {
                formattedNumberProps[key] = `"${value}"`;
            } else {
                formattedNumberProps[key] = `{${value}}`;
            }
        });

        const formattedNumberPropsString = Object.entries(formattedNumberProps)
            .map(([key, value]) => {
                if (typeof value === "undefined") {
                    return null;
                }
                return `${key}=${value}`;
            })
            .join(" ");

        return {
            textContent: `typeof ${rowValue} === "undefined" || ${rowValue} === null ? ${emptyText} : <FormattedNumber ${formattedNumberPropsString} />`,
        };
    }

    if (textConfig.type === "staticSelect") {
        const labelsVariableName = `${textConfig.field}Labels`;

        const labelMapping = textConfig.values
            .map((valueOption) => {
                const value = typeof valueOption === "string" ? valueOption : valueOption.value;
                const label = typeof valueOption === "string" ? valueOption : valueOption.label;
                return `${value}: ${getFormattedMessageNode(`${messageIdPrefix}.${value}`, label)}`;
            })
            .join(", ");

        const labelMappingVar = `const ${labelsVariableName}: Record<string, ReactNode> = { ${labelMapping} };`;
        const textContent = `(${rowValue} == null ? ${emptyText} : ${labelsVariableName}[` + `\`\${${rowValue}}\`` + `] ?? ${rowValue})`;

        return {
            textContent,
            variableDefinitions: [labelMappingVar],
        };
    }

    return {
        textContent: stringTextContent,
    };
};

export const getCombinationColumnRenderCell = (column: GridCombinationColumnConfig<string>, messageIdPrefix: string) => {
    const gridCellContentProps: Record<string, string> = {};
    const allVariableDefinitions: string[] = [];

    if (column.primaryText) {
        const { textContent, variableDefinitions = [] } = getTextForCellContent(column.primaryText, `${messageIdPrefix}.primaryText`);
        gridCellContentProps.primaryText = textContent;
        allVariableDefinitions.push(...variableDefinitions);
    }

    if (column.secondaryText) {
        const { textContent, variableDefinitions = [] } = getTextForCellContent(column.secondaryText, `${messageIdPrefix}.secondaryText`);
        gridCellContentProps.secondaryText = textContent;
        allVariableDefinitions.push(...variableDefinitions);
    }

    const allUniqueVariableDefinitions = Array.from(new Set(allVariableDefinitions));

    return `({ row }) => {
        ${allUniqueVariableDefinitions.join("\n")}
        return <GridCellContent ${Object.entries(gridCellContentProps)
            .map(([key, value]) => `${key}={${value}}`)
            .join(" ")} />;
    }`;
};

const getFieldNamesFromText = (textConfig: Field<string>): string[] => {
    if (typeof textConfig === "string") {
        return [textConfig];
    }

    if (textConfig.type === "static") {
        return [];
    }

    if (textConfig.type === "formattedMessage") {
        return Object.values(textConfig.valueFields).flatMap((value) => getFieldNamesFromText(value));
    }

    return [textConfig.field];
};

export const getAllColumnFieldNames = (column: GridCombinationColumnConfig<string>): string[] => {
    const fieldNames: string[] = [];

    if (column.primaryText) {
        fieldNames.push(...getFieldNamesFromText(column.primaryText));
    }

    if (column.secondaryText) {
        fieldNames.push(...getFieldNamesFromText(column.secondaryText));
    }

    return fieldNames.filter((fieldName) => fieldName !== "id");
};
