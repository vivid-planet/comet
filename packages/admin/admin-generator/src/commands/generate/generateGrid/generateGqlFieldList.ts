import objectPath from "object-path";

import { type ActionsGridColumnConfig, type GridColumnConfig } from "../generate-command";
import { getAllColumnFieldNames, type GridCombinationColumnConfig } from "./combinationColumn";

type FieldsObjectType = { [key: string]: FieldsObjectType | boolean | string };
const recursiveStringify = (obj: FieldsObjectType): string => {
    let ret = "";
    let prefixField = "";
    for (const key in obj) {
        const valueForKey = obj[key];
        if (typeof valueForKey === "boolean") {
            ret += `${prefixField}${key}`;
        } else if (typeof valueForKey === "string") {
            ret += `${prefixField}${key}${valueForKey}`;
        } else {
            ret += `${prefixField}${key} { ${recursiveStringify(valueForKey)} }`;
        }
        prefixField = " ";
    }
    return ret;
};

export function generateGqlFieldList({
    columns,
}: {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    columns: Array<GridColumnConfig<any> | GridCombinationColumnConfig<string> | ActionsGridColumnConfig>;
}) {
    const fieldsObject: FieldsObjectType = columns.reduce<FieldsObjectType>((acc, field) => {
        if (field.type !== "actions") {
            if (field.type === "combination") {
                getAllColumnFieldNames(field).map((fieldName) => {
                    objectPath.set(acc, fieldName, true);
                });
            } else {
                objectPath.set(acc, field.name, true);
            }
        }
        return acc;
    }, {});
    return recursiveStringify(fieldsObject);
}
