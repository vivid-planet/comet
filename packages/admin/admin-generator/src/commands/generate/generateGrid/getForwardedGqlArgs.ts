import { type IntrospectionField, type IntrospectionInputValue } from "graphql";

import { type Imports } from "../utils/generateImportsCode";
import { type Prop } from "./generateGrid";

export type GqlArg = { type: string; name: string; queryOrMutationName: string };

export function getForwardedGqlArgs(gqlFields: IntrospectionField[]) {
    const ret: {
        imports: Imports;
        prop: Prop;
        gqlArg: GqlArg;
    }[] = [];

    const supportedGqlArgs = ["offset", "limit", "sort", "search", "filter", "input"]; // this arguments need to be handled differently or are already handled somewhere else

    getArgs(gqlFields, supportedGqlArgs).forEach((arg) => {
        let prop: Prop;
        const imports: Imports = [];

        if (arg.type === "ID" || arg.type === "String" || arg.type === "DateTime") {
            prop = { name: arg.name, optional: false, type: "string" };
        } else if (arg.type === "Boolean") {
            prop = { name: arg.name, optional: false, type: "boolean" };
        } else if (arg.type === "Int" || arg.type === "Float") {
            prop = { name: arg.name, optional: false, type: "number" };
        } else if (arg.type === "JSONObject") {
            prop = { name: arg.name, optional: false, type: "unknown" };
        } else {
            prop = { name: arg.name, optional: false, type: arg.type };
            imports.push({ name: arg.type, importPath: "@src/graphql.generated" });
        }

        ret.push({ gqlArg: { name: arg.name, type: arg.type, queryOrMutationName: arg.gqlField.name }, prop, imports });
    });

    return ret;
}

function getArgs(gqlFields: IntrospectionField[], skipGqlArgs: string[]) {
    return gqlFields.reduce<{ name: string; type: string; gqlArg: IntrospectionInputValue; gqlField: IntrospectionField }[]>((acc, gqlField) => {
        gqlField.args.forEach((gqlArg) => {
            if (skipGqlArgs.includes(gqlArg.name)) return acc;
            if (gqlArg.type.kind !== "NON_NULL" || gqlArg.defaultValue) return acc;

            const gqlType = gqlArg.type.ofType;

            let type = "unknown";
            if (gqlType.kind === "SCALAR") {
                const nativeScalars = ["ID", "String", "Boolean", "Int", "Float", "DateTime", "JSONObject"];
                if (!nativeScalars.includes(gqlType.name)) {
                    // probably just add to gqlFields-array, but needs to be tested
                    console.warn(`Currently not supported special SCALAR of type ${gqlType.name} in param ${gqlArg.name} of ${gqlField.name}`);
                } else {
                    type = gqlType.name;
                }
            } else if (gqlType.kind === "INPUT_OBJECT") {
                type = gqlType.name;
            } else {
                throw new Error(`Not supported kind ${gqlType.kind}, arg: ${gqlArg.name}`);
            }

            acc.push({ name: gqlArg.name, type, gqlArg, gqlField });
        });
        return acc;
    }, []);
}
