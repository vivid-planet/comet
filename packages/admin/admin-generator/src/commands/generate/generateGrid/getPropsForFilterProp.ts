import { type IntrospectionInputObjectType } from "graphql";

import { type GridConfig } from "../generate-command";
import { type Imports } from "../utils/generateImportsCode";
import { type Prop } from "./generateGrid";

export function getPropsForFilterProp<T extends { __typename?: string }>({
    config,
    filterType,
}: {
    config: GridConfig<T>;
    filterType: IntrospectionInputObjectType;
}): {
    hasFilterProp: boolean;
    imports: Imports;
    props: Prop[];
} {
    if (!config.filterProp) return { hasFilterProp: false, imports: [], props: [] };

    const filterTypeName = `GQL${filterType.name}`;

    return {
        hasFilterProp: true,
        imports: [{ name: filterTypeName, importPath: "@src/graphql.generated" }],
        props: [{ name: "filter", optional: true, type: filterTypeName }],
    };
}
