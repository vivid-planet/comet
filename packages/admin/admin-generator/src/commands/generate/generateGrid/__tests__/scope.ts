import { buildSchema, introspectionFromSchema } from "graphql";

import { type GridConfig } from "../../generate-command";
import { parseSource } from "../../utils/test-helper";
import { generateGrid } from "../generateGrid";

describe("Grid Scope", () => {
    const schema = buildSchema(`
        type Query {
            products(scope: ProductScope!): [Product!]!
        }

        type Mutation {
            updateProduct(id: ID!, input: ProductInput!): Product!
        }

        input ProductScope {
            domain: String!
        }

        input ProductInput {
            title: String
        }

        type Product {
            id: ID!
            title: String
        }
    `);
    type GQLProduct = {
        __typename?: "Product";
        id: string;
        title: string;
    };
    const introspection = introspectionFromSchema(schema);

    it("uses scope from context by default", async () => {
        const gridConfig: GridConfig<GQLProduct> = {
            type: "grid",
            gqlType: "Product",
            columns: [
                {
                    type: "text",
                    name: "title",
                },
            ],
        };

        const formOutput = generateGrid(
            {
                gqlIntrospection: introspection,
                baseOutputFilename: "ProductsGrid",
                exportName: "ProductsGrid",
                targetDirectory: "/test",
            },
            gridConfig,
        );
        expect(formOutput.code).toMatchSnapshot();
    });
    it("generates prop for scope", async () => {
        const gridConfig: GridConfig<GQLProduct> = {
            type: "grid",
            gqlType: "Product",
            columns: [
                {
                    type: "text",
                    name: "title",
                },
            ],
            scopeAsProp: true,
        };

        const formOutput = generateGrid(
            {
                gqlIntrospection: introspection,
                baseOutputFilename: "ProductsGrid",
                exportName: "ProductsGrid",
                targetDirectory: "/test",
            },
            gridConfig,
        );
        expect(formOutput.code).toMatchSnapshot();
    });

    it("generates correct gql import for scope", async () => {
        const gridConfig: GridConfig<GQLProduct> = {
            type: "grid",
            gqlType: "Product",
            columns: [
                {
                    type: "text",
                    name: "title",
                },
            ],
            scopeAsProp: true,
        };

        const formOutput = generateGrid(
            {
                gqlIntrospection: introspection,
                baseOutputFilename: "ProductsGrid",
                exportName: "ProductsGrid",
                targetDirectory: "/test",
            },
            gridConfig,
        );

        const source = parseSource(formOutput.code);
        const importDeclarations = source.getImportDeclarations();

        const graphqlImportDeclaration = importDeclarations.find(
            (getImportDeclaration) => getImportDeclaration.getModuleSpecifierValue() === "@src/graphql.generated",
        );
        expect(graphqlImportDeclaration).toBeDefined();

        const gqlScopeImport = graphqlImportDeclaration?.getNamedImports().find((namedImport) => namedImport.getName() === "GQLProductScope");
        expect(gqlScopeImport).toBeDefined();
    });
});
