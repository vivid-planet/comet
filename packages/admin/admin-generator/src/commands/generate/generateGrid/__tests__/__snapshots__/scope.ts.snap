// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid Scope generates prop for scope 1`] = `
"import {
        GQLProductsGridQuery,
        GQLProductsGridQueryVariables,
        GQLProductsFormFragment,
        GQLCreateProductMutation,
        GQLCreateProductMutationVariables,
        GQLUpdateProductPositionMutation,
        GQLUpdateProductPositionMutationVariables,
        GQLDeleteProductMutation,
        GQLDeleteProductMutationVariables
    } from "./ProductsGrid.generated";
    import { FormattedMessage } from "react-intl";
import { FormattedNumber } from "react-intl";
import { useIntl } from "react-intl";
import { ReactNode } from "react";
import { gql } from "@apollo/client";
import { useApolloClient } from "@apollo/client";
import { useQuery } from "@apollo/client";
import { Button } from "@comet/admin";
import { CrudContextMenu } from "@comet/admin";
import { CrudMoreActionsMenu } from "@comet/admin";
import { DataGridToolbar } from "@comet/admin";
import { ExportApi } from "@comet/admin";
import { filterByFragment } from "@comet/admin";
import { GridFilterButton } from "@comet/admin";
import { GridCellContent } from "@comet/admin";
import { GridColDef } from "@comet/admin";
import { dataGridDateTimeColumn } from "@comet/admin";
import { dataGridDateColumn } from "@comet/admin";
import { dataGridIdColumn } from "@comet/admin";
import { dataGridManyToManyColumn } from "@comet/admin";
import { dataGridOneToManyColumn } from "@comet/admin";
import { renderStaticSelectCell } from "@comet/admin";
import { messages } from "@comet/admin";
import { muiGridFilterToGql } from "@comet/admin";
import { muiGridSortToGql } from "@comet/admin";
import { StackLink } from "@comet/admin";
import { FillSpace } from "@comet/admin";
import { Tooltip } from "@comet/admin";
import { useBufferedRowCount } from "@comet/admin";
import { useDataGridExcelExport } from "@comet/admin";
import { useDataGridRemote } from "@comet/admin";
import { usePersistentColumnState } from "@comet/admin";
import { BlockPreviewContent } from "@comet/cms-admin";
import { Alert } from "@mui/material";
import { Box } from "@mui/material";
import { IconButton } from "@mui/material";
import { Typography } from "@mui/material";
import { useTheme } from "@mui/material";
import { Menu } from "@mui/material";
import { MenuItem } from "@mui/material";
import { ListItemIcon } from "@mui/material";
import { ListItemText } from "@mui/material";
import { CircularProgress } from "@mui/material";
import { DataGridPro } from "@mui/x-data-grid-pro";
import { GridLogicOperator } from "@mui/x-data-grid-pro";
import { GridRenderCellParams } from "@mui/x-data-grid-pro";
import { GridSlotsComponent } from "@mui/x-data-grid-pro";
import { GridToolbarProps } from "@mui/x-data-grid-pro";
import { GridColumnHeaderTitle } from "@mui/x-data-grid-pro";
import { GridToolbarQuickFilter } from "@mui/x-data-grid-pro";
import { GridRowOrderChangeParams } from "@mui/x-data-grid-pro";
import { ProductScope } from "@src/graphql.generated";
import { Add as AddIcon } from "@comet/admin-icons";
import { Edit as EditIcon } from "@comet/admin-icons";
import { Info as InfoIcon } from "@comet/admin-icons";
import { Excel as ExcelIcon } from "@comet/admin-icons";

    const productsFragment = gql\`
        fragment ProductsForm on Product {
            id
            title
        }
    \`;

    const productsQuery = gql\`
    query ProductsGrid($scope: ProductScope!, $offset: Int!, $limit: Int!) {
        products(scope: $scope, offset: $offset, limit: $limit) {
            nodes { ...ProductsForm } totalCount
        }
    }
    \${productsFragment}
    \`;

    

    


    
    function ProductsGridToolbar() {
        return (
            <DataGridToolbar>
                <FillSpace />
        <Button responsive startIcon={<AddIcon />} component={StackLink} pageName="add" payload="add">
        <FormattedMessage id="product.productsForm.newEntry" defaultMessage={\`New Product\`}  />
    </Button>
            </DataGridToolbar>
        );
    }

    type Props = {
            scope: ProductScope;
        };

    export function ProductsGrid({scope }: Props) {
        
        const intl = useIntl();
        const dataGridProps = { ...useDataGridRemote(), ...usePersistentColumnState("ProductsGrid") };
        
        

        

        const columns: GridColDef<GQLProductsFormFragment>[] = [
            {field: "title",
headerName: intl.formatMessage({ id: "product.title", defaultMessage: "Title" }),
filterable: false,
sortable: false,
flex: 1,
minWidth: 150,},
                {field: "actions",
headerName: "",
sortable: false,
filterable: false,
type: "actions",
align: "right",
pinned: "right",
width: 52,
renderCell: (params) => {
                            return (
                                <>
                                
                                        <IconButton color="primary" component={StackLink} pageName="edit" payload={params.row.id}>
                                            <EditIcon />
                                        </IconButton>
                                </>
                            );
                                },}
        ];

        

        const { data, loading, error } = useQuery<GQLProductsGridQuery, GQLProductsGridQueryVariables>(productsQuery, {
            variables: {
                scope, offset: dataGridProps.paginationModel.page * dataGridProps.paginationModel.pageSize, limit: dataGridProps.paginationModel.pageSize, sort: muiGridSortToGql(dataGridProps.sortModel, columns)
            },
        });
        const rowCount = useBufferedRowCount(data?.products.totalCount);
        if (error) throw error;
        const rows = data?.products.nodes ?? [];

        

        return (
            <DataGridPro
                {...dataGridProps}
                rows={rows}
                rowCount={rowCount}
                columns={columns}
                loading={loading}
                slots={{
                                toolbar: ProductsGridToolbar as GridSlotsComponent["toolbar"],
                            }}
                            
                
            />
        );
    }
    "
`;

exports[`Grid Scope uses scope from context by default 1`] = `
"import {
        GQLProductsGridQuery,
        GQLProductsGridQueryVariables,
        GQLProductsFormFragment,
        GQLCreateProductMutation,
        GQLCreateProductMutationVariables,
        GQLUpdateProductPositionMutation,
        GQLUpdateProductPositionMutationVariables,
        GQLDeleteProductMutation,
        GQLDeleteProductMutationVariables
    } from "./ProductsGrid.generated";
    import { FormattedMessage } from "react-intl";
import { FormattedNumber } from "react-intl";
import { useIntl } from "react-intl";
import { ReactNode } from "react";
import { gql } from "@apollo/client";
import { useApolloClient } from "@apollo/client";
import { useQuery } from "@apollo/client";
import { Button } from "@comet/admin";
import { CrudContextMenu } from "@comet/admin";
import { CrudMoreActionsMenu } from "@comet/admin";
import { DataGridToolbar } from "@comet/admin";
import { ExportApi } from "@comet/admin";
import { filterByFragment } from "@comet/admin";
import { GridFilterButton } from "@comet/admin";
import { GridCellContent } from "@comet/admin";
import { GridColDef } from "@comet/admin";
import { dataGridDateTimeColumn } from "@comet/admin";
import { dataGridDateColumn } from "@comet/admin";
import { dataGridIdColumn } from "@comet/admin";
import { dataGridManyToManyColumn } from "@comet/admin";
import { dataGridOneToManyColumn } from "@comet/admin";
import { renderStaticSelectCell } from "@comet/admin";
import { messages } from "@comet/admin";
import { muiGridFilterToGql } from "@comet/admin";
import { muiGridSortToGql } from "@comet/admin";
import { StackLink } from "@comet/admin";
import { FillSpace } from "@comet/admin";
import { Tooltip } from "@comet/admin";
import { useBufferedRowCount } from "@comet/admin";
import { useDataGridExcelExport } from "@comet/admin";
import { useDataGridRemote } from "@comet/admin";
import { usePersistentColumnState } from "@comet/admin";
import { BlockPreviewContent } from "@comet/cms-admin";
import { Alert } from "@mui/material";
import { Box } from "@mui/material";
import { IconButton } from "@mui/material";
import { Typography } from "@mui/material";
import { useTheme } from "@mui/material";
import { Menu } from "@mui/material";
import { MenuItem } from "@mui/material";
import { ListItemIcon } from "@mui/material";
import { ListItemText } from "@mui/material";
import { CircularProgress } from "@mui/material";
import { DataGridPro } from "@mui/x-data-grid-pro";
import { GridLogicOperator } from "@mui/x-data-grid-pro";
import { GridRenderCellParams } from "@mui/x-data-grid-pro";
import { GridSlotsComponent } from "@mui/x-data-grid-pro";
import { GridToolbarProps } from "@mui/x-data-grid-pro";
import { GridColumnHeaderTitle } from "@mui/x-data-grid-pro";
import { GridToolbarQuickFilter } from "@mui/x-data-grid-pro";
import { GridRowOrderChangeParams } from "@mui/x-data-grid-pro";
import { ProductScope } from "@src/graphql.generated";
import { useContentScope } from "@comet/cms-admin";
import { Add as AddIcon } from "@comet/admin-icons";
import { Edit as EditIcon } from "@comet/admin-icons";
import { Info as InfoIcon } from "@comet/admin-icons";
import { Excel as ExcelIcon } from "@comet/admin-icons";

    const productsFragment = gql\`
        fragment ProductsForm on Product {
            id
            title
        }
    \`;

    const productsQuery = gql\`
    query ProductsGrid($offset: Int!, $limit: Int!, $scope: ProductContentScopeInput!) {
        products(offset: $offset, limit: $limit, scope: $scope) {
            nodes { ...ProductsForm } totalCount
        }
    }
    \${productsFragment}
    \`;

    

    


    
    function ProductsGridToolbar() {
        return (
            <DataGridToolbar>
                <FillSpace />
        <Button responsive startIcon={<AddIcon />} component={StackLink} pageName="add" payload="add">
        <FormattedMessage id="product.productsForm.newEntry" defaultMessage={\`New Product\`}  />
    </Button>
            </DataGridToolbar>
        );
    }

    

    export function ProductsGrid() {
        
        const intl = useIntl();
        const dataGridProps = { ...useDataGridRemote(), ...usePersistentColumnState("ProductsGrid") };
        const { scope } = useContentScope();
        

        

        const columns: GridColDef<GQLProductsFormFragment>[] = [
            {field: "title",
headerName: intl.formatMessage({ id: "product.title", defaultMessage: "Title" }),
filterable: false,
sortable: false,
flex: 1,
minWidth: 150,},
                {field: "actions",
headerName: "",
sortable: false,
filterable: false,
type: "actions",
align: "right",
pinned: "right",
width: 52,
renderCell: (params) => {
                            return (
                                <>
                                
                                        <IconButton color="primary" component={StackLink} pageName="edit" payload={params.row.id}>
                                            <EditIcon />
                                        </IconButton>
                                </>
                            );
                                },}
        ];

        

        const { data, loading, error } = useQuery<GQLProductsGridQuery, GQLProductsGridQueryVariables>(productsQuery, {
            variables: {
                scope, offset: dataGridProps.paginationModel.page * dataGridProps.paginationModel.pageSize, limit: dataGridProps.paginationModel.pageSize, sort: muiGridSortToGql(dataGridProps.sortModel, columns)
            },
        });
        const rowCount = useBufferedRowCount(data?.products.totalCount);
        if (error) throw error;
        const rows = data?.products.nodes ?? [];

        

        return (
            <DataGridPro
                {...dataGridProps}
                rows={rows}
                rowCount={rowCount}
                columns={columns}
                loading={loading}
                slots={{
                                toolbar: ProductsGridToolbar as GridSlotsComponent["toolbar"],
                            }}
                            
                
            />
        );
    }
    "
`;
